ledger/src/main/java/com/scalar/dl/ledger/database/scalardb/AbstractScalarNamespaceRegistry.java

```diff
@@ -0,0 +1,185 @@
+  @Nullable private final DistributedTransactionAdmin transactionAdmin;
+  private final Set<TableMetadataProvider> tableMetadataProviders;
+  private final ImmutableMap<String, TableMetadata> storageTables;
+  private final ImmutableMap<String, TableMetadata> transactionTables;
+  private final Retry retry;
+
+  @SuppressFBWarnings("EI_EXPOSE_REP2")
+  public AbstractScalarNamespaceRegistry(
+      ServerConfig config,
+      DistributedStorage storage,
+      DistributedStorageAdmin storageAdmin,
+      DistributedTransactionAdmin transactionAdmin,
+      Set<TableMetadataProvider> tableMetadataProviders) {
+    this.config = checkNotNull(config);
+    this.storage = checkNotNull(storage);
+    this.storageAdmin = checkNotNull(storageAdmin);
+    this.transactionAdmin = transactionAdmin;
+    this.tableMetadataProviders = checkNotNull(tableMetadataProviders);
+    this.storageTables = collectStorageTables();
+    this.transactionTables = collectTransactionTables();
+    this.retry =
+        Retry.of(
+            "retry",
+            RetryConfig.custom()
+                .maxAttempts(MAX_ATTEMPTS)
+                .intervalFunction(IntervalFunction.ofExponentialBackoff(SLEEP_BASE_MILLIS, 2.0))
+                .retryOnException(
+                    e -> {
+                      LOGGER.warn("Namespace creation failed");
+                      return e instanceof IllegalArgumentException
+                          && (e.getMessage().startsWith(CoreError.NAMESPACE_NOT_FOUND.buildCode())
+                              || e.getMessage().startsWith(CoreError.TABLE_NOT_FOUND.buildCode()));
+                    })
+                .build());
+  }
+
+  public AbstractScalarNamespaceRegistry(
+      ServerConfig config,
+      DistributedStorage storage,
+      DistributedStorageAdmin storageAdmin,
+      Set<TableMetadataProvider> tableMetadataProviders) {
+    this(config, storage, storageAdmin, null, tableMetadataProviders);
+  }
+
+  @Override
+  public void create(String namespace) {
+    Retry.decorateRunnable(
+            retry,
+            () -> {
+              createNamespaceManagementTable();
+              createNamespace(namespace);
+              addNamespaceEntry(namespace);
+            })
+        .run();
+  }
+
+  private void createNamespaceManagementTable() {
+    try {
+      storageAdmin.createTable(
+          config.getNamespace(), NAMESPACE_TABLE_NAME, NAMESPACE_TABLE_METADATA, true);
+    } catch (ExecutionException e) {
+      throw new DatabaseException(CommonError.CREATING_NAMESPACE_TABLE_FAILED, e, e.getMessage());
+    }
+  }
+
+  private void createNamespace(String namespace) {
+    try {
+      String fullNamespaceName = config.getNamespace() + NAMESPACE_NAME_SEPARATOR + namespace;
+      storageAdmin.createNamespace(fullNamespaceName, true);
+      // this sleep can be removed after negative cache is disabled in ScalarDB
+      Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
```

Comment:
Currently, ScalarDB caches an empty result (negative cache) when a specified table is not found (e.g., due to the metadata sync delay mentioned above). It takes 60 seconds by default to expire, and users cannot create the namespace during that time. So, we would like to avoid it.

But, after the negative cache is disabled in ScalarDB, this delay/sleep can be achieved through the retry logic. Since sleep here is unnecessary and inefficient for other underlying storage, remove it after the fixed ScalarDB version is released.

Reply:
3.17 is released now, so it can be updated?